lib/withHotKeys.js

```diff
@@ -0,0 +1,76 @@
+import React, {PureComponent} from 'react';
+import HotKeys from './HotKeys';
+
+/**
+ * withHotKeys is an HOC that provides the wrappedComponent with the ability to implement keyboard actions
+ * without the user wrapping every component with a <HotKeys> component individually
+ *
+ * See examples/master/HOCWrappedNode.js for an example implementation
+ * Follow the steps below to use the HOC:
+ *
+ * @example <caption>Example usage of withHotKeys.</caption>
+ * // Returns the HOC-wrapped component.
+ * // 1. Declared a key map that with the actionName as key and keyboardKeys as values
+ * const ACTION_KEY_MAP = {
+ *     'logConsole' : 'down',
+ * };
+ *
+ * class BasicBox extends React.Component {
+ *
+ * // 2. declare 'hotKeyHandlers' within the Component's class definition
+ *   hotKeyHandlers: {
+ *     'logConsole': this.logConsole.bind(this),
+ *   }
+ *
+ *   logConsole() {
+ *     console.log('a hotkey is pressed');
+ *   }
+ *
+ *   render() {
+ *     return (
+ *         <div tabIndex="0">
+ *             Press the down arrow
+ *         </div>
+ *     );
+ *   }
+ * }
+ *
+ * // 3. Wrap the Component with withHotKeys
+ * export default withHotKeys(ACTION_KEY_MAP)(BasicBox);
+ * @returns {function} Returns the HOC-wrapped component.
+ *
+ * @param {Object} keyMap an action-to-keyboard-key mapping
+ * @summary An HOC that provides the wrappedComponent with the ability to implement keyboard actions
+ */
+const withHotKeys = (keyMap, handlers) => ((Component) =>
+    class withHotKeysWrapper extends PureComponent {
+      static propTypes = Component.propTypes;
+      state = {
+        handlers: {}
+      };
+
+      componentDidMount() {
+        this.setState({handlers: this._ref.hotKeyHandlers});
```

Comment:
This is much better than the first version. Thanks for taking the time to rework it.

I still have mixed feelings about this - reaching into the children of `withHotKeysWrapper` to pick off the `hotKeyHandlers` attribute feels a little like information flow in the wrong direction. I guess I am still having trouble understanding a use case for this. 

From what I can see, this HoC can be used to create a component that you can pass arbitrary (but necessarily singular) children to, who define their handlers later on (but they can't change after mounting without causing `withHotKeysWrapper` to update, somehow). So the actions stay the same - they are defined and encapsulated in the `withHotKeysWrapper` component, but you can re-use that component and mount it with different children, who define different handlers.

So if you were only to use that `withHotKeysWrapper` component once, I would say that it would be more declarative (and avoid a double render) to just use one `HotKeys` component at the top of the application that defines the `keyMap` and many different `HotKeys` descendants who define (different) `handlers`.

However, if you want to use it in many places in your application to enforce a sort of interface (children of the `withHotKeysWrapper` component can/should define handlers to the sequences in the `withHotKeysWrapper `'s`keyMap`), then I can perhaps begin to understand there being a use case for this. But I am not sure if such a use case can be better handled slightly differently.

In my head I am thinking of it as this:

```
class MyComponent extends Component {
  render() {
    return (
      <HotKeys handlers={ myHandlers } >
        //...
      </HotKeys>
    )
  }
}

class MyContainer extends Component {
  render() {
    return {
      <HotKeys keyMap={myKeyMap} >
        //-----
        // ...|- <MyComponent />
        // ......|- <MyComponent />
        // .....|- <MyComponent />
      </HotKeys>
    }
  }
}
```

Being replaced by this:

```
class MyComponent extends Component {
  constructor(props, context) {
    super(props, context)

    this.hotKeyHandlers = myHandlers
  }

  render() {
    return (
      //...
    )
  }
}

const MyContainer = withHotKeys(myKeyMap)(MyComponent)
```

which does remove the reference to `HotKeys` in `MyComponent`, but it comes at the cost of - I think- double rendering as the `withHotKeysWrapper` component must wait to have its child to get the handlers, to re-render that child with the information that it already has.

What do you think?

Reply:
I understand the concerns of 1) getting the child's state via ref and 2) one extra render on `componentDidMount`. I think that these are trade-offs for 1) abstracting out `<HotKeys>` from the `Component` render and 2) sectioning handlers into `hotKeyHandlers` instead of defining in `render` (which could already be fill with other logic). Modularizing hotkeys support is advantageous for managing multiple components with different needs for keyboard support. With that said, if there's a way for the HOC to not re-render but also not duplicate `HotKeys` functionalities, please let me know! You have more expertise :)

Reply:
I can't think of a way to provide a HOC that matches the functionality of the component (allowing `handlers` to be defined at mount-time), so this is the best solution available to us at the moment.